futures-select-macro/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+//! The futures-rs `select! macro implementation.
+
+#![recursion_limit="128"]
+
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use proc_macro2::Span;
+use proc_macro_hack::proc_macro_hack;
+use quote::quote;
+use syn::{parenthesized, parse_quote, Expr, Ident, Pat, Token};
+use syn::parse::{Parse, ParseStream};
+
+mod kw {
+    syn::custom_keyword!(complete);
+    syn::custom_keyword!(futures_crate_path);
+}
+
+struct Select {
+    futures_crate_path: Option<syn::Path>,
+    // span of `complete`, then expression after `=> ...`
+    complete: Option<Expr>,
+    default: Option<Expr>,
+    normal_fut_exprs: Vec<Expr>,
+    normal_fut_handlers: Vec<(Pat, Expr)>,
+}
+
+#[allow(clippy::large_enum_variant)]
+enum CaseKind {
+    Complete,
+    Default,
+    Normal(Pat, Expr),
+}
+
+impl Parse for Select {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let mut select = Select {
+            futures_crate_path: None,
+            complete: None,
+            default: None,
+            normal_fut_exprs: vec![],
+            normal_fut_handlers: vec![],
+        };
+
+        // When `futures_crate_path(::path::to::futures::lib)` is provided,
+        // it sets the path through which futures library functions will be
+        // accessed.
+        if input.peek(kw::futures_crate_path) {
+            input.parse::<kw::futures_crate_path>()?;
+            let content;
+            parenthesized!(content in input);
+            select.futures_crate_path = Some(content.parse()?);
+        }
+
+        while !input.is_empty() {
+            let case_kind = if input.peek(kw::complete) {
+                // `complete`
+                if select.complete.is_some() {
+                    return Err(input.error("multiple `complete` cases found, only one allowed"));
+                }
+                input.parse::<kw::complete>()?;
+                CaseKind::Complete
+            } else if input.peek(Token![default]) {
```

Comment:
What's the reason that this doesn't look exactly like the `complete` branch?

Reply:
Syn has a token for `default` because it is used as a keyword in `default fn` syntax, while `complete` is not a keyword.